crates/vm/rust/src/contract.rs

```diff
@@ -348,16 +383,60 @@ where
             after_tx_fn: self.after_tx_fn,
             bank_execute_fn: self.bank_execute_fn,
             bank_query_fn: self.bank_query_fn,
+            withhold_fee_fn: self.withhold_fee_fn,
+            finalize_fee_fn: Some(finalize_fee_fn),
             cron_execute_fn: self.cron_execute_fn,
-        }));
+        }
+    }
+
+    pub fn with_cron_execute<E13A>(
+        self,
+        cron_execute_fn: CronExecuteFn<E13A>,
+    ) -> ContractBuilder<M1, E1, M2, M3, M5, M6, E2, E3, E4, E5, E6, E7, E8, E9, E10, E11, E12, E13A>
+    {
+        ContractBuilder {
+            instantiate_fn: self.instantiate_fn,
+            execute_fn: self.execute_fn,
+            migrate_fn: self.migrate_fn,
+            receive_fn: self.receive_fn,
+            reply_fn: self.reply_fn,
+            query_fn: self.query_fn,
+            before_tx_fn: self.before_tx_fn,
+            after_tx_fn: self.after_tx_fn,
+            bank_execute_fn: self.bank_execute_fn,
+            bank_query_fn: self.bank_query_fn,
+            withhold_fee_fn: self.withhold_fee_fn,
+            finalize_fee_fn: self.finalize_fee_fn,
+            cron_execute_fn: Some(cron_execute_fn),
+        }
+    }
+
+    pub fn build(self) -> ContractWrapper {
+        let index = CONTRACTS
+            .get_or_init(Default::default)
+            .push_get_index(Box::new(ContractImpl {
```

Comment:
Another bug found: previously, what we did was:

```rust
let contracts = CONTRACTS.get_or_init(Default::default);
let index = contracts.len();
contracts.push(ContractImpl { /* ... */ });
```

This does work with multithreading, because between the time we do `len` and `push`, another thread might have inserted an element, causing the index to be incorrect.

The correct way is to use `push_get_index` to push a new element and get its index in one go atomically.